File: leetcode/cpp/valid_parentheses/valid_parentheses__stack.cpp
#include <stack>
#include <string>

using namespace std;

class ValidParentheses_Stack
{
  private:
    static bool arePair(char open, char close)
    {
        return (open == '(' && close == ')') || (open == '{' && close == '}') || (open == '[' && close == ']');
    }

  public:
    static bool isValid(string s)
    {

        stack<char> stack;

        for (int i = 0; i < s.length(); i++)
        {
            if (s[i] == '(' || s[i] == '{' || s[i] == '[')
            {
                stack.push(s[i]);
            }
            else if (s[i] == ')' || s[i] == '}' || s[i] == ']')
            {
                if (stack.empty() || !arePair(stack.top(), s[i]))
                {
                    return false;
                }
                else
                {
                    stack.pop();
                }
            }
        }

        return stack.empty() ? true : false;
    }
};

File: leetcode/csharp/valid_parentheses/valid_parentheses__stack.cs
using System.Collections.Generic;

namespace LeetCode.Solutions.q20_valid_parentheses;

public class ValidParentheses_Stack
{
    public static bool Solution(string parentheses)
    {
        Stack<char> stack = new();

        for (int current = 0; current < parentheses.Length; current++)
        {
            if (IsEmpty(stack))
            {
                if (IsCloseParenthesis(parentheses[current]))
                {
                    return false;
                }
                else
                {
                    stack.Push(parentheses[current]);
                }
            }
            else
            {
                if (ArePair(stack.Peek(), parentheses[current]))
                {
                    stack.Pop();
                }
                else
                {
                    stack.Push(parentheses[current]);
                }
            }
        }

        return IsEmpty(stack);
    }

    private static bool IsEmpty(Stack<char> stack)
    {
        return stack.Count == 0;
    }

    private static bool IsCloseParenthesis(char parenthesis)
    {
        return parenthesis == ')' || parenthesis == ']' || parenthesis == '}';
    }

    private static bool ArePair(char open, char close)
    {
        return open == '(' && close == ')' ||
                open == '[' && close == ']' ||
                open == '{' && close == '}';
    }
}
